go/teams/seitan_v2.go

```diff
@@ -0,0 +1,276 @@
+package teams
+
+import (
+	"bytes"
+	"fmt"
+	"strings"
+
+	"crypto/hmac"
+	"crypto/rand"
+	"crypto/sha512"
+	"encoding/base64"
+	"errors"
+
+	"github.com/keybase/client/go/libkb"
+	"github.com/keybase/go-crypto/ed25519"
+	"golang.org/x/crypto/nacl/secretbox"
+	"golang.org/x/net/context"
+
+	keybase1 "github.com/keybase/client/go/protocol/keybase1"
+)
+
+// Seitan tokens v2 have a '+' as the sixth character. We use this
+// to distinguish from email invite tokens (and team names).
+// See `IsSeitanyV2`
+const seitanEncodedIKeyV2PlusOffset = 6
+
+// "Invite Key Version 2"
+type SeitanIKeyV2 string
+
+// "Packed Encrypted Invite Key Version 2"
+// All following 3 structs should be considered one. When any changes,
+// Version in PEPubKey has to be bumped up.
+type SeitanPEPubKey struct {
+	_struct              bool `codec:",toarray"`
+	Version              uint
+	TeamKeyGeneration    keybase1.PerTeamKeyGeneration
+	RandomNonce          keybase1.BoxNonce
+	EncryptedKeyAndLabel []byte // keybase1.SeitanPubKeyAndLabel MsgPacked and encrypted
+}
+
+func GenerateIKeyV2() (ikey SeitanIKeyV2, err error) {
+	buf, err := generateIKey(seitanEncodedIKeyV2PlusOffset)
+	if err != nil {
+		return ikey, err
+	}
+	return SeitanIKeyV2(string(buf)), err
+}
+
+// IsSeitany is a very conservative check of whether a given string looks
+// like a Seitan token. We want to err on the side of considering strings
+// Seitan tokens, since we don't mistakenly want to send botched Seitan
+// tokens to the server.
+func IsSeitanyV2(s string) bool {
+	return len(s) > seitanEncodedIKeyV2PlusOffset && strings.IndexByte(s, '+') > 1
+}
+
+// GenerateIKeyV2FromString safely creates SeitanIKey value from
+// plaintext string. Only format is checked - any 18-character token
+// with '+' character at position 6 can be "Invite Key". Alphabet is
+// not checked, as it is only a hint for token generation and it can
+// change over time, but we assume that token length stays the same.
+func GenerateIKeyV2FromString(token string) (ikey SeitanIKeyV2, err error) {
+	if len(token) != SeitanEncodedIKeyLength {
+		return ikey, fmt.Errorf("invalid token length: expected %d characters, got %d", SeitanEncodedIKeyLength, len(token))
+	}
+	if token[seitanEncodedIKeyV2PlusOffset] != '+' {
+		return ikey, fmt.Errorf("invalid token format: expected %dth character to be '+'", seitanEncodedIKeyV2PlusOffset+1)
+	}
+
+	return SeitanIKeyV2(strings.ToLower(token)), nil
+}
+
+func (ikey SeitanIKeyV2) String() string {
+	return strings.ToLower(string(ikey))
+}
+
+// "Stretched Invite Key"
+type SeitanSIKeyV2 [SeitanScryptKeylen]byte
+
+func (ikey SeitanIKeyV2) GenerateSIKey() (sikey SeitanSIKeyV2, err error) {
+	sikeyv1, err := SeitanIKey(ikey).GenerateSIKey()
```

Comment:
I'd rather have a generic stretching function that takes a buffer instead of this casting. And than use that function in both v1 and v2.

Reply:
agreed, i mixed refactoring this yesterday!

Reply:
done